Parent/Parent/ParentAppDelegate.swift

```diff
@@ -118,27 +92,132 @@ class ParentAppDelegate: UIResponder, UIApplicationDelegate {
     @objc func routeToRemindable(from response: UNNotificationResponse) {
         let userInfo = response.notification.request.content.userInfo
         if let urlString = userInfo[RemindableActionURLKey] as? String, let url = URL(string: urlString) {
-            Router.sharedInstance.route(visibleController, toURL: url, modal: true)
+            Router.sharedInstance.route(topMostViewController()!, toURL: url, modal: true)
+        }
+    }
+
+    func setup(session: KeychainEntry) {
+        environment.userDidLogin(session: session)
+        CoreWebView.keepCookieAlive(for: environment)
+        // UX requires that students are given color schemes in a specific order.
+        // The method call below ensures that we always start with the first color scheme.
+        ColorCoordinator.clearColorSchemeDictionary()
+        if Locale.current.regionCode != "CA" {
+            let crashlyticsUserId = "\(session.userID)@\(session.baseURL.host ?? session.baseURL.absoluteString)"
+            Crashlytics.sharedInstance().setUserIdentifier(crashlyticsUserId)
+        }
+        // Legacy CanvasKeymaster support
+        let legacyClient = CKIClient(baseURL: session.baseURL, token: session.accessToken)!
+        legacyClient.actAsUserID = session.actAsUserID
+        legacyClient.originalIDOfMasqueradingUser = session.originalUserID
+        legacyClient.originalBaseURL = session.originalBaseURL
+        legacyClient.fetchCurrentUser().subscribeNext { user in
+            legacyClient.setValue(user, forKey: "currentUser")
```

Comment:
can we not use `legacyClient.currentUser`? Worried about type safety (though I'm not _that_ worried)

Reply:
It's not publicly settable. I could just make it a public property.